lib/Sema/DerivedConformances.cpp

```diff
@@ -23,7 +23,87 @@
 using namespace swift;
 using namespace DerivedConformance;
 
-ValueDecl *DerivedConformance::getDerivableRequirement(NominalTypeDecl *nominal,
+bool DerivedConformance::derivesProtocolConformance(TypeChecker &tc,
+                                                    NominalTypeDecl *nominal,
+                                                    ProtocolDecl *protocol) {
+  // Only known protocols can be derived.
+  auto knownProtocol = protocol->getKnownProtocolKind();
+  if (!knownProtocol)
+    return false;
+
+  if (auto *enumDecl = dyn_cast<EnumDecl>(nominal)) {
+    switch (*knownProtocol) {
+        // The presence of a raw type is an explicit declaration that
+        // the compiler should derive a RawRepresentable conformance.
+      case KnownProtocolKind::RawRepresentable:
+        return enumDecl->hasRawType();
+
+        // Enums without associated values can implicitly derive Equatable and
+        // Hashable conformance.
+      case KnownProtocolKind::Equatable:
+        return canDeriveEquatable(tc, enumDecl, protocol);
+      case KnownProtocolKind::Hashable:
+        return canDeriveHashable(tc, enumDecl, protocol);
+
+        // @objc enums can explicitly derive their _BridgedNSError conformance.
+      case KnownProtocolKind::BridgedNSError:
+        return enumDecl->isObjC() && enumDecl->hasCases()
+            && enumDecl->hasOnlyCasesWithoutAssociatedValues();
+
+        // Enums without associated values and enums with a raw type of String
+        // or Int can explicitly derive CodingKey conformance.
+      case KnownProtocolKind::CodingKey: {
+        Type rawType = enumDecl->getRawType();
+        if (rawType) {
+          auto parentDC = enumDecl->getDeclContext();
+          ASTContext &C = parentDC->getASTContext();
+
+          auto nominal = rawType->getAnyNominal();
+          return nominal == C.getStringDecl() || nominal == C.getIntDecl();
+        }
+
+        // hasOnlyCasesWithoutAssociatedValues will return true for empty enums;
+        // empty enumas are allowed to conform as well.
+        return enumDecl->hasOnlyCasesWithoutAssociatedValues();
+      }
+
+      default:
+        return false;
+    }
+  } else if (isa<StructDecl>(nominal) || isa<ClassDecl>(nominal)) {
+    // Structs and classes can explicitly derive Encodable and Decodable
+    // conformance (explicitly meaning we can synthesize an implementation if
+    // a type conforms manually).
+    if (*knownProtocol == KnownProtocolKind::Encodable ||
+        *knownProtocol == KnownProtocolKind::Decodable) {
+      // FIXME: This is not actually correct. We cannot promise to always
```

Comment:
But you're passing in a type checker instance to this function, no?

Reply:
That's true now. This code was moved wholesale from `ASTContext` to `DerivedConformances` at the recommendation that it belonged in Sema. Since IIRC this is originally @itaiferber's code/comment relating to `Codable`, I didn't want to touch anything outside the scope of `Equatable`/`Hashable`.

Reply:
It's fine to clean up other code (and in fact encouraged) when implementing new features.

Reply:
Good to know for the future! If you're interested in merging this as-is (referencing your comment below), then I'm happy to take a look at that cleanup in another PR.